index/segment/postings_list.go

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package segment
+
+import (
+	"sync"
+
+	"github.com/RoaringBitmap/roaring"
+)
+
+// roaringPostingsList wraps a roaring.Bitmap w/ a mutex for thread safety.
+type roaringPostingsList struct {
+	sync.RWMutex
+	bitmap *roaring.Bitmap
+}
+
+// NewPostingsList returns a new PostingsList.
+func NewPostingsList() PostingsList {
+	return &roaringPostingsList{
+		bitmap: roaring.NewBitmap(),
+	}
+}
+
+func (d *roaringPostingsList) Insert(i DocID) {
+	d.Lock()
+	d.bitmap.Add(uint32(i))
+	d.Unlock()
+}
+
+func (d *roaringPostingsList) Intersect(other ImmutablePostingsList) {
+	o, ok := other.(*roaringPostingsList)
+	if !ok {
+		panic("Intersect only supported between roaringDocId sets")
```

Comment:
Hm I'd be in preference of bubbling this all up as errors tbh.  I know this is an application bug that ideally would be compile time enforced but because of the interfaces used here cannot be - having said that I could definitely imagine in the future needing to return an error for intersect/difference/etc so it'd probably better to just add error as a type to these interface methods and bubble up the errors instead.

Otherwise we'll start to use this as a pattern in other places or take it as "acceptable."  

Everything we write should try as much as possible to avoid causing panics themselves and bubble up errors so callers can make the call on whether an operation causing an error is fatal enough to panic.

Reply:
Fair. I do think panics have a place (e.g. before allowing corrupt data to be written to disk), this isn't one of those cases. Will bubble error everywhere.

Reply:
done.